core/include/vecmem/containers/device_vector.hpp

```diff
@@ -118,6 +124,60 @@ namespace vecmem {
 
       /// @}
 
+      /// @name Payload modification functions
+      /// @{
+
+      /// Assign new values to the vector (not thread-safe)
+      VECMEM_HOST_AND_DEVICE
+      void assign( size_type count, const_reference value );
+      /// Assign new values to the vector (not thread-safe)
+      template< typename InputIt,
+                std::enable_if_t<
+                   details::is_iterator_of< InputIt, value_type >::value,
+                   bool > = true >
+      VECMEM_HOST_AND_DEVICE
+      void assign( InputIt other_begin, InputIt other_end ) {
+
+         // This can only be done on a resizable vector.
+         assert( m_size != nullptr );
+
+         // Remove all previous elements.
+         clear();
+
+         // Create copies of all of the elements one-by-one. It's very
+         // inefficient, but we can't make any assumptions about the type of the
+         // input iterator received by this function.
+         atomic< size_type > asize( m_size );
+         for( InputIt itr = other_begin; itr != other_end; ++itr ) {
+            construct( asize.fetch_add( 1 ), *itr );
+         }
+      }
+
+      /// Add a new element at the end of the vector ("kind of" thread-safe)
+      template< typename... Args >
+      VECMEM_HOST_AND_DEVICE
+      reference emplace_back( Args&&... args );
+      /// Add a new element at the end of the vector ("kind of" thread-safe)
```

Comment:
Is it obvious what "kind of" means? Thread-safe unless capacity needs increased?

Reply:
As long as every thread is only making use of `push_back(...)` and/or `emplace_back(...)`, that should be thread-safe. But if any other thread is trying to use any of the non-thread-safe functions at the same time, then these will also not work correctly.

I guess that does fall under thread safe still. :thinking: Since on any class thread-safe and non-thread-safe functions would have that behaviour... So the wording here could indeed be changed. Do you have a proposal?

Reply:
Yeah, probably just "thread safe".

But what about the capacity? Is it fixed in the context in which we consider this thread safe?

Reply:
The only way in which I imagine that we can have resizable vectors is by constructing "vector buffers" in the right way.

The memory allocated by `vecmem::vector` or `vecmem::jagged_vector` (so `std::vector`...) can not be allowed to be "resized" by device code. As propagating that information back to the host vector object would be a **very** fragile operation. (You can't just call `std::vector::resize(...)` on the host object after the kernel finished running, since that would overwrite the memory that the kernel filled up just previously.) It wouldn't be impossible, but it would require very fragile code...

So you would create, in this PR's example, a `vecmem::data::vector_buffer<T>` with a separate "size" and "capacity" parameter. The `vecmem::device_vector<T>` created on top of that buffer will now know that it can change the "size" of the vector, but only up to the buffer's "capacity".

So yes, the capacity is decided in device code, ahead of running the kernel. This is just how device code works. :frowning: You must put restrictions / assumptions around your memory management, the algorithms are not free to use however much memory they want. (If you ever tried programming in Fortran, this is pretty similar to how memory management there works / worked.)

Now... One thing that I really hate in the current code, and which I was very curious if either of you would notice, is how `vecmem::data::vector_buffer<T>` is set up to be resizable. It's a very misleading interface in this PR. :frowning: You should have a look, and then we can discuss afterwards how it could be made better...

Reply:
Okay, removed the `"kind of"`. :wink: